In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import statsmodels.api as sm
%matplotlib inline
pd.options.display.max_rows = 999
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [2]:
from ast import literal_eval
def clean(x):
    return literal_eval(x)

In [3]:
accre_jobs_piped = pd.read_csv('../data/accre_jobs_cleaned.csv', converters={'node_numbers': clean})

In [6]:
accre_jobs_piped.shape

(3730198, 17)

# Maggi -- 
    Question: 3 & 4   Calculation of % of actual memory used and mean % for each group

## Notes on conversion of Mn to Mc
For Mn, convert to Mc as follows: digits / (cpus/nodes)<br>
For Mc, use the digits as is<br>
usedmem is stated in Mn, but designated as M or possibly with no designation<br>
recmem can be stated in either Mc, Mn, or none<br>
Question: The documentation says that cpus is the total cpus for the job (i.e., not the cpus per node);
    however, some jobs have 1 cpu and 2 nodes, 1 cpu and 51 nodes; so, I don't understand that.<br>
Make two new columns in memory_use reqmem_mc and usedmem_mc

In [7]:
#display the data set
accre_jobs_piped.head(2)

jobid  account      user    reqmem    usedmem      reqtime     usedtime  \
0  15925210  treviso  arabella  122880Mn  65973.49M  13-18:00:00  13-18:00:28   
1  15861126  treviso  arabella  122880Mn  67181.12M  13-18:00:00  12-14:50:56   

   nodes  cpus   partition exitcode      state    nodelist  req_total_seconds  \
0      1    24  production      0:0  COMPLETED  ['cn1531']            1188000   
1      1    24  production      0:0  COMPLETED  ['cn1441']            1188000   

   used_total_seconds  reqmem_mc   usedmem_mc  
0             1188028     5120.0  2748.895417  
1             1090256     5120.0  2799.213333

In [8]:
#look at our groupings of 'state'
accre_jobs_piped['state'].value_counts()
#we will just want to use 'COMPLETED'

COMPLETED              3718637
CANCELLED                 6064
RUNNING                   3105
FAILED                    1379
CANCELLED by 686562        681
OUT_OF_MEMORY               94
TIMEOUT                     80
CANCELLED by 505355         63
CANCELLED by 855431         17
CANCELLED by 782611         13
CANCELLED by 397600          7
CANCELLED by 0               5
CANCELLED by 486541          4
CANCELLED by 200557          4
CANCELLED by 454080          4
CANCELLED by 199766          4
CANCELLED by 649319          3
CANCELLED by 9202            3
CANCELLED by 9201            3
CANCELLED by 483348          3
CANCELLED by 791651          2
CANCELLED by 546080          2
CANCELLED by 199066          2
CANCELLED by 895426          2
CANCELLED by 124006          2
CANCELLED by 515423          2
CANCELLED by 693461          2
CANCELLED by 666860          2
CANCELLED by 90423           2
CANCELLED by 781109          1
CANCELLED by 785271          1
CANCELLED by 651701          1
CANCELLE

In [9]:
#look at our groupings of 'partition'
accre_jobs_piped['partition'].value_counts()
#we will just want to use 'production'

production    3311788
nogpfs         327652
pascal          48004
turing          39406
maxwell          3348
Name: partition, dtype: int64

In [10]:
#look at 'exitcode' = 0:0 in 'production' by 'state'
accre_jobs_piped[(accre_jobs_piped['exitcode']=='0:0') & (accre_jobs_piped['partition']=='production')]['state'].value_counts()
#definitely limit to completed based on the 'cancelled*' and 'running' entries

COMPLETED              3302522
CANCELLED                 4449
RUNNING                   2859
CANCELLED by 686562        680
TIMEOUT                     79
CANCELLED by 9201            2
Name: state, dtype: int64

In [11]:
#pull out the initial slice of data for memory usage analysis
memory_use = accre_jobs_piped[(accre_jobs_piped['partition']=='production') & (accre_jobs_piped['exitcode']=='0:0')
                              & (accre_jobs_piped['state']=='COMPLETED')]

In [12]:
memory_use.head(2)

jobid  account      user    reqmem    usedmem      reqtime     usedtime  \
0  15925210  treviso  arabella  122880Mn  65973.49M  13-18:00:00  13-18:00:28   
1  15861126  treviso  arabella  122880Mn  67181.12M  13-18:00:00  12-14:50:56   

   nodes  cpus   partition exitcode      state    nodelist  req_total_seconds  \
0      1    24  production      0:0  COMPLETED  ['cn1531']            1188000   
1      1    24  production      0:0  COMPLETED  ['cn1441']            1188000   

   used_total_seconds  reqmem_mc   usedmem_mc  
0             1188028     5120.0  2748.895417  
1             1090256     5120.0  2799.213333

In [13]:
memory_use.shape

(3302522, 17)

In [15]:
#examples where reqmem in Mc and nodes > 1
memory_use[(memory_use['reqmem'].str.contains('Mc')) & (memory_use['nodes']>1)][:10]

jobid account     user   reqmem    usedmem     reqtime  usedtime  \
13240   17105760  summer    buddy  20480Mc   6453.04M    04:40:00  00:00:56   
13257   17105819  summer    buddy  20480Mc   7587.56M    04:40:00  00:11:43   
13342   17105966  summer    buddy  20480Mc   7587.15M    04:40:00  00:14:03   
26744   17125144  summer    buddy  20480Mc  20479.96M    04:40:00  00:17:50   
27044   17125676  summer    buddy  20480Mc  20463.48M    04:40:00  00:34:25   
28241   17126611  summer    buddy  20480Mc   6712.42M    04:40:00  00:03:37   
61606   17173216    arra  needham   8192Mc  65535.89M    06:00:00  00:06:44   
118478  17255274   belle  osbaldo   8192Mc          0  1-00:00:00  00:01:15   
160741  17324222  crowns     tena   5120Mc   9651.66M  2-00:00:00  16:47:47   
164606  17329000   round   shenna   8000Mc          0    10:00:00  00:00:12   

        nodes  cpus   partition exitcode      state  \
13240       2     1  production      0:0  COMPLETED   
13257       2     1  production      0:0  COMPLETED   
13342       2     1  production      0:0  COMPLETED   
26744       3     1  production      0:0  COMPLETED   
27044       3     1  production      0:0  COMPLETED   
28241       3     1  production      0:0  COMPLETED   
61606      19     8  production      0:0  COMPLETED   
118478      4    12  production      0:0  COMPLETED   
160741     37     1  production      0:0  COMPLETED   
164606      2     1  production      0:0  COMPLETED   

                                                 nodelist  req_total_seconds  \
13240                                ['cn1434', 'cn1435']              16800   
13257                                ['cn1434', 'cn1435']              16800   
13342                                ['cn1434', 'cn1435']              16800   
26744                      ['cn1448', 'cn1449', 'cn1450']              16800   
27044                      ['cn1465', 'cn1466', 'cn1467']              16800   
28241                      ['cn1430', 'cn1431', 'cn1432']              16800   
61606   ['cn1281', 'cn1283', 'cn1297', 'cn1298', 'cn13...              21600   
118478           ['cn1293', 'cn1302', 'cn1355', 'cn1356']              86400   
160741  ['cn1094', 'cn1125', 'cn1266', 'cn1277', 'cn12...             172800   
164606                               ['cn1291', 'cn1292']              36000   

        used_total_seconds  reqmem_mc    usedmem_mc  
13240                   56    20480.0   12906.08000  
13257                  703    20480.0   15175.12000  
13342                  843    20480.0   15174.30000  
26744                 1070    20480.0   61439.88000  
27044                 2065    20480.0   61390.44000  
28241                  217    20480.0   20137.26000  
61606                  404     8192.0  155647.73875  
118478                  75     8192.0       0.00000  
160741               60467     5120.0  357111.42000  
164606                  12     8000.0       0.00000

In [16]:
#it looks like there are some jobs with 0 usedmem, does that make sense?
memory_use['usedmem_mc'].value_counts()
#no it does not, these don't seem valid for evaluating efficiency

0.0000       537283
1.5600        31241
0.1975        30085
1.5575        29904
6.2300        16918
              ...  
3493.0900         1
1416.2400         1
1393.7020         1
2559.6200         1
4122.6600         1
Name: usedmem_mc, Length: 884478, dtype: int64

In [17]:
#here's what the data looks like
memory_use[memory_use['usedmem_mc'] ==0]['used_total_seconds'].describe().apply(lambda x: format(x, 'f'))
#most are not long running jobs, but some in the top 25% are

count     537283.000000
mean          37.988003
std         2058.355733
min            0.000000
25%            7.000000
50%           12.000000
75%           20.000000
max      1013546.000000
Name: used_total_seconds, dtype: object

In [18]:
#since we'll be calculating ratios to look at efficiency, we don't want any rows where reqmem or usedmem is 0
#also remove a few stray rows where cpus = 0
memory_use = memory_use[(memory_use['reqmem_mc']>0) & (memory_use['usedmem_mc']>0) & (memory_use['cpus']>0)]

In [19]:
#some jobs have fewer cpus than nodes, which doesn't make sense
check_cpus = memory_use[memory_use['cpus']<memory_use['nodes']][['account','cpus','nodes','reqmem', 'usedmem', 'reqmem_mc','usedmem_mc' ]]

In [20]:
#here's what that data looks like
check_cpus
#it's probably not valid for efficiency ratios

account  cpus  nodes   reqmem  usedmem      reqmem_mc   usedmem_mc
3453     winged     1      3   4096Mn  570.32M   12288.000000  1710.960000
3454     winged     1      4   4096Mn  669.61M   16384.000000  2678.440000
3457     winged     1      3   4096Mn  620.31M   12288.000000  1860.930000
3460     winged     1      4   4096Mn  629.38M   16384.000000  2517.520000
3464     winged     1      3   4096Mn  641.92M   12288.000000  1925.760000
...         ...   ...    ...      ...      ...            ...          ...
3626272   burro     3     20  51200Mn   53.06M  341333.333333   353.733333
3629669   burro     3     20  51200Mn  235.39M  341333.333333  1569.266667
3669106   burro     3     20  51200Mn  102.50M  341333.333333   683.333333
3669107   burro     3     20  51200Mn   99.59M  341333.333333   663.933333
3670648   burro     3     20  51200Mn  105.52M  341333.333333   703.466667

[1192 rows x 7 columns]

In [21]:
#is it limited to particular accounts?
check_cpus['account'].value_counts()
#not particularly

winged          294
chickpeas       293
angelys         196
sharlin         186
wood             62
retailer         52
burro            35
mignonette       15
crowns           13
arra             11
summer            6
peanuts           6
casaba            6
carrot            4
crab              3
grisette          3
blackberries      3
delblush          1
cms_lowprio       1
horseradish       1
clara             1
Name: account, dtype: int64

In [22]:
memory_use['account'].value_counts()

cep              464990
cms              452991
summer           372681
carrot           179941
galia            172315
cms_lowprio      116401
casaba           104134
portabella       103329
bunch             88829
orange            76909
bulk              49270
horseradish       44361
sharlin           43898
sweet             38711
key               34944
mixed             34099
retailer          32049
boysenberries     27957
wood              26226
round             23557
cmsadmin          23421
atemoyas          22084
garlic            19329
wax               17287
alkmene           15532
grisette          15176
celery            14510
winged            14108
sunburst          12876
fresh             11025
chickpeas         10486
regal              9641
large              9019
texas              8606
minnewashta        6555
clara              6324
belle              5687
lychees            5454
st                 5335
muscat             4774
bertanne           4405
seeded          

In [23]:
#remove the rows where cpus < nodes
memory_use = memory_use[memory_use['cpus']>=memory_use['nodes']]

In [24]:
memory_use['used_hours'] = (memory_use['used_total_seconds'] / 3600)

In [25]:
#memory weighting formula
#full_df['MEM_WEIGHT'] = full_df['MEM_PERCENT_USED']*full_df['TOTAL_SEC_USED']
#full_df.groupby('ACCOUNT')['MEM_WEIGHT'].sum()/full_df.groupby('ACCOUNT')['TOTAL_SEC_USED'].sum()
#alkmene 38.351490
#almonds 46.785636
#amaranth 4.135943

In [26]:
memory_use['avg_mem_use'] = memory_use['usedmem_mc'] / memory_use['reqmem_mc'] * 100

In [27]:
memory_use['mem_weight'] = memory_use['avg_mem_use'] * memory_use['used_total_seconds']

In [28]:
memory_use.groupby('account')['mem_weight'].sum() / memory_use.groupby('account')['used_total_seconds'].sum()

account
alkmene          38.398614
almonds          21.713587
amaranth          4.136165
angelys           2.701473
anise            50.589604
antares          67.351935
arra             61.596102
atemoyas         14.127690
baby             14.528333
baking           42.547142
bay               5.710036
bell              6.252304
belle            24.432158
bertanne         25.736394
bibb             68.479948
blackberries     30.977459
bon              19.198852
boysenberries    12.872584
brazilnuts       33.754403
broccoli         49.645722
bulk             46.143998
bunch            15.861109
burro            43.340971
cactus           33.614457
cantared         19.378149
carambola        21.421409
carrot           45.486137
casaba           29.247738
celery           15.567764
cep              33.662324
chasselas        66.380879
chayote           1.515647
chickpeas        34.243716
chipilin         46.867197
cinnabar         37.181179
clara            12.245560
cms              40.

In [29]:
memory_use.sort_values('used_total_seconds', ascending=False)

jobid        account     user    reqmem     usedmem      reqtime  \
1173511  19005552  sugranineteen    daron    4096Mn     544.95M  12-12:00:00   
1173522  19005634  sugranineteen    daron    4096Mn     556.69M  12-12:00:00   
1173500  19005517  sugranineteen    daron    4096Mn     591.11M  12-12:00:00   
1173501  19005518  sugranineteen    daron    4096Mn     589.16M  12-12:00:00   
1173544  19005936  sugranineteen    daron    4096Mn     544.98M  12-12:00:00   
...           ...            ...      ...       ...         ...          ...   
3537368  23797646        alkmene  abigail   47104Mn     104.74M   1-00:00:00   
291869   17500485         baking  christa  184320Mn     475.46M   1-00:09:00   
232476   17412039         baking  christa  184320Mn  184319.70M   1-00:09:00   
3423542  23564776        alkmene  abigail   47104Mn    1569.34M   1-00:00:00   
330091   17557834         baking  christa  184320Mn  184319.60M   2-06:09:00   

            usedtime  nodes  cpus   partition exitcode      state    nodelist  \
1173511  19-20:04:25      1     1  production      0:0  COMPLETED  ['cn1135']   
1173522  19-19:53:24      1     1  production      0:0  COMPLETED  ['cn1126']   
1173500  19-15:04:26      1     1  production      0:0  COMPLETED  ['cn1088']   
1173501  19-15:04:18      1     1  production      0:0  COMPLETED  ['cn1124']   
1173544  19-14:48:21      1     1  production      0:0  COMPLETED  ['cn1131']   
...              ...    ...   ...         ...      ...        ...         ...   
3537368     00:00:00      1     6  production      0:0  COMPLETED  ['cn1266']   
291869      00:00:00      1    12  production      0:0  COMPLETED  ['cn1326']   
232476      00:00:00      1    12  production      0:0  COMPLETED  ['cn1509']   
3423542     00:00:00      1     6  production      0:0  COMPLETED  ['cn1386']   
330091      00:00:00      1    12  production      0:0  COMPLETED  ['cn1262']   

         req_total_seconds  used_total_seconds     reqmem_mc    usedmem_mc  \
1173511            1080000             1713865   4096.000000    544.950000   
1173522            1080000             1713204   4096.000000    556.690000   
1173500            1080000             1695866   4096.000000    591.110000   
1173501            1080000             1695858   4096.000000    589.160000   
1173544            1080000             1694901   4096.000000    544.980000   
...                    ...                 ...           ...           ...   
3537368              86400                   0   7850.666667     17.456667   
291869               86940                   0  15360.000000     39.621667   
232476               86940                   0  15360.000000  15359.975000   
3423542              86400                   0   7850.666667    261.556667   
330091              194940                   0  15360.000000  15359.966667   

         used_hours  avg_mem_use    mem_weight  
1173511  476.073611    13.304443  2.280202e+07  
1173522  475.890000    13.591064  2.328427e+07  
1173500  471.073889    14.431396  2.447371e+07  
1173501  471.071667    14.383789  2.439286e+07  
1173544  470.805833    13.305176  2.255096e+07  
...             ...          ...           ...  
3537368    0.000000     0.222359  0.000000e+00  
291869     0.000000     0.257954  0.000000e+00  
232476     0.000000    99.999837  0.000000e+00  
3423542    0.000000     3.331649  0.000000e+00  
330091     0.000000    99.999783  0.000000e+00  

[2764033 rows x 20 columns]

# Plot avg mem usage

In [30]:
avg_memuse_byacct = memory_use.groupby(['account'])[['reqmem_mc', 'usedmem_mc', 'used_hours']].sum().reset_index()

In [31]:
#add a column for the unweighted average
avg_memuse_byacct['avg_usage'] = np.round(avg_memuse_byacct['usedmem_mc'] / avg_memuse_byacct['reqmem_mc'] * 100, decimals=2)

In [32]:
#add a column for average memory usage quartile
avg_memuse_byacct['mem_quartile'] = pd.qcut(avg_memuse_byacct['avg_usage'], q = 4, labels=['Q1','Q2','Q3','Q4'])

In [33]:
#add a column for used_hours quartile
avg_memuse_byacct['hours_quartile'] = pd.qcut(avg_memuse_byacct['used_hours'], q = 4, 
                            labels=['Q1','Q2','Q3','Q4'])

In [34]:
#add a column for used_hours category to be used in plots
avg_memuse_byacct['total_time_ctgry'] = pd.qcut(avg_memuse_byacct['used_hours'], q = 4,
                            labels=['Ttl Time: Low', 'Ttl Time: MedLo','Ttl Time: MedHi','Ttl Time: High Usage'])

In [35]:
#add a column for used_hours category to be used in table
avg_memuse_byacct['ttl_time_usg_ctgry'] = pd.qcut(avg_memuse_byacct['used_hours'], q = 4,
                            labels=['1Low','2MedLo','3MedHi','4High'])

In [36]:
#add a column for avg_usage category to be used in table
avg_memuse_byacct['avg_mem_usg_ctgry'] = pd.qcut(avg_memuse_byacct['avg_usage'], q = 4, labels=['1Low','2MedLo','3MedHi','4High'])

In [37]:
avg_memuse_byacct.sort_values('avg_usage', ascending=False)

account     reqmem_mc    usedmem_mc    used_hours  avg_usage  \
75          medium  1.517670e+07  1.046097e+07  1.397343e+03      68.93   
76      mignonette  2.252800e+05  1.542592e+05  3.034556e+01      68.47   
113   strawberries  7.549013e+06  5.079764e+06  5.658872e+02      67.29   
5          antares  3.655680e+05  2.426081e+05  1.171669e+02      66.36   
109          shell  4.505600e+04  2.782108e+04  7.206506e+02      61.75   
110          shiny  1.686923e+06  1.005946e+06  5.390319e+02      59.63   
60          jicama  1.158724e+07  6.767345e+06  6.157806e+03      58.40   
120          texas  3.668187e+07  2.112502e+07  2.085284e+04      57.59   
6             arra  5.558272e+05  3.031670e+05  1.559681e+02      54.54   
107       shallots  3.246080e+06  1.748024e+06  2.935500e+01      53.85   
101        roasted  1.081685e+06  5.798736e+05  1.953978e+02      53.61   
116         summer  4.446754e+09  2.340071e+09  2.743177e+05      52.62   
30       chasselas  1.308515e+07  6.807766e+06  1.057141e+04      52.03   
91          pinova  4.565432e+06  2.364135e+06  1.372695e+03      51.78   
43        delblush  8.934400e+05  4.625195e+05  6.306188e+03      51.77   
126            wax  7.080755e+07  3.578591e+07  2.269833e+05      50.54   
114        sturmer  1.213722e+07  6.045134e+06  1.518130e+04      49.81   
33        chipilin  8.846760e+05  4.247904e+05  8.493763e+03      48.02   
105         seeded  4.358516e+07  2.004801e+07  4.622674e+03      46.00   
100       retailer  3.450146e+08  1.525321e+08  1.023420e+05      44.21   
22           burro  1.695132e+06  7.466824e+05  7.003799e+03      44.05   
65         kumquat  1.297920e+05  5.700368e+04  6.006417e+01      43.92   
19        broccoli  3.258368e+05  1.428123e+05  1.274166e+03      43.83   
88          pepino  1.438066e+06  5.765411e+05  2.667717e+03      40.09   
38        cmsadmin  4.933222e+07  1.948796e+07  1.845737e+03      39.50   
61       josephine  3.141120e+05  1.207763e+05  1.990586e+03      38.45   
57     horseradish  3.916414e+08  1.488890e+08  3.823565e+04      38.02   
119          sweet  4.101023e+08  1.531543e+08  4.001220e+05      37.35   
102          rocha  3.308800e+04  1.228172e+04  6.898278e+01      37.12   
118      sunflower  2.827764e+06  1.041762e+06  1.047978e+04      36.84   
26          carrot  1.167186e+09  4.131155e+08  1.129283e+06      35.39   
4            anise  7.154238e+06  2.499509e+06  4.034354e+03      34.94   
49           fresh  1.065443e+08  3.685065e+07  1.976532e+05      34.59   
82            nino  7.000000e+03  2.385344e+03  6.131389e+00      34.08   
44            dill  4.377600e+05  1.490328e+05  3.646956e+02      34.04   
124        treviso  2.711040e+06  9.201500e+05  2.089144e+04      33.94   
0          alkmene  1.793044e+07  6.048471e+06  9.883721e+03      33.73   
79           mixed  2.135494e+08  7.061518e+07  6.018411e+04      33.07   
78     minnewashta  1.125294e+08  3.623371e+07  1.855505e+05      32.20   
130             yu  8.837180e+05  2.821726e+05  5.754917e+01      31.93   
29             cep  1.884784e+09  5.920948e+08  2.871719e+06      31.41   
121       thomcord  2.195456e+06  6.842214e+05  1.065313e+04      31.17   
12           belle  7.248348e+07  2.238292e+07  2.039781e+04      30.88   
13        bertanne  3.543766e+07  1.082719e+07  2.654792e+04      30.55   
23          cactus  1.380352e+06  4.136377e+05  8.772911e+02      29.97   
68           lemon  2.160640e+05  6.454760e+04  4.777322e+02      29.87   
9           baking  6.439371e+06  1.873608e+06  1.474915e+04      29.10   
34        cinnabar  2.467840e+05  7.009681e+04  8.390333e+01      28.40   
128         winged  3.557869e+07  9.971923e+06  1.236047e+05      28.03   
64          kinnow  2.299221e+06  6.353317e+05  4.448297e+03      27.63   
52          garlic  3.095839e+08  8.344585e+07  5.030999e+05      26.95   
37     cms_lowprio  6.816290e+08  1.756128e+08  8.708560e+05      25.76   
51           galia  1.428246e+09  3.

In [38]:
avg_memuse_byacct['avg_usage'].describe().apply(lambda x: format(x, 'f'))

count    131.000000
mean      23.542290
std       18.179492
min        0.180000
25%        8.015000
50%       19.240000
75%       34.335000
max       68.930000
Name: avg_usage, dtype: object

In [75]:
q1cut = avg_memuse_byacct['avg_usage'].quantile(.25)
q2cut = avg_memuse_byacct['avg_usage'].quantile(.5)
q3cut = avg_memuse_byacct['avg_usage'].quantile(.75)
maxuse =  avg_memuse_byacct['avg_usage'].max()
q1_subplot_title = 'Low Efficiency - 0% to ' + str(int(q1cut)) + '%'
q2_subplot_title = 'MedLo Efficiency - ' + str(int(q1cut)) + '% to ' + str(int(q2cut)) + '%'
q3_subplot_title = 'MedHi Efficiency - ' + str(int(q2cut)) + '% to ' + str(int(q3cut)) + '%'
q4_subplot_title = 'High Efficiency - ' + str(int(q3cut)) + '% to ' + str(int(maxuse)) + '%'
q4_subplot_title


'High Efficiency - 34% to 68%'

In [76]:
avg_memuse_byacct['used_hours'].describe().apply(lambda x: format(x, 'f'))

count        131.000000
mean      105861.801957
std       394025.731354
min            0.032500
25%          207.885000
50%         2667.717500
75%        26197.583333
max      3049115.967501
Name: used_hours, dtype: object

In [77]:
avg_memuse_byacct[['used_hours', 'account']].sort_values('used_hours',ascending = False)

used_hours        account
36   3.049116e+06            cms
29   2.871719e+06            cep
26   1.129283e+06         carrot
37   8.708560e+05    cms_lowprio
129  7.183964e+05           wood
51   5.396737e+05          galia
99   5.056481e+05          regal
52   5.030999e+05         garlic
97   4.739416e+05     portabella
119  4.001220e+05          sweet
116  2.743177e+05         summer
126  2.269833e+05            wax
27   2.126984e+05         casaba
49   1.976532e+05          fresh
78   1.855505e+05    minnewashta
63   1.437247e+05            key
108  1.294916e+05        sharlin
128  1.236047e+05         winged
72   1.207671e+05        lychees
100  1.023420e+05       retailer
17   9.742938e+04  boysenberries
21   8.892657e+04          bunch
20   6.885571e+04           bulk
32   6.226926e+04      chickpeas
117  6.135802e+04       sunburst
79   6.018411e+04          mixed
83   5.966085e+04         orange
115  4.440307e+04  sugranineteen
57   3.823565e+04    horseradish
41   3.617897e+04     conference
24   3.367976e+04       cantared
104  2.841185e+04          round
13   2.654792e+04       bertanne
2    2.584725e+04       amaranth
85   2.549680e+04        pasilla
54   2.264360e+04        harovin
124  2.089144e+04        treviso
120  2.085284e+04          texas
12   2.039781e+04          belle
35   1.725799e+04          clara
15   1.591016e+04   blackberries
7    1.538765e+04       atemoyas
114  1.518130e+04        sturmer
9    1.474915e+04         baking
121  1.065313e+04       thomcord
30   1.057141e+04      chasselas
118  1.047978e+04      sunflower
0    9.883721e+03        alkmene
33   8.493763e+03       chipilin
66   8.086753e+03          large
22   7.003799e+03          burro
42   6.409696e+03           crab
43   6.306188e+03       delblush
60   6.157806e+03         jicama
18   5.987541e+03     brazilnuts
81   5.963148e+03         muscat
112  5.854606e+03             st
105  4.622674e+03         seeded
64   4.448297e+03         kinnow
127  4.208659e+03          white
67   4.162068e+03         leaves
4    4.034354e+03          anise
8    3.943189e+03           baby
94   3.605920e+03       plantain
53   3.423569e+03       grisette
88   2.667717e+03         pepino
16   2.448097e+03            bon
71   2.325375e+03           long
106  2.044074e+03        serrano
61   1.990586e+03      josephine
38   1.845737e+03       cmsadmin
25   1.524140e+03      carambola
75   1.397343e+03         medium
91   1.372695e+03         pinova
19   1.274166e+03       broccoli
56   1.056836e+03           horn
45   8.803522e+02         emmons
23   8.772911e+02         cactus
10   7.851433e+02            bay
109  7.206506e+02          shell
70   7.106503e+02          lollo
86   6.737503e+02        peanuts
113  5.658872e+02   strawberries
46   5.507714e+02      fairchild
110  5.390319e+02          shiny
40   5.187575e+02        concord
11   5.087653e+02           bell
1    4.818278e+02        almonds
47   4.793081e+02     fiddlehead
68   4.777322e+02          lemon
14   4.516033e+02           bibb
84   3.960528e+02      oroblanco
44   3.646956e+02           dill
58   3.498511e+02        huaguan
28   3.489789e+02         celery
93   3.336161e+02         pitted
111  2.895122e+02         spring
77   2.203722e+02          milwa
101  1.953978e+02        roasted
90   1.920172e+02           pink
125  1.885925e+02          water
6    1.559681e+02           arra
59   1.466336e+02         ingrid
50   1.442156e+02       galangal
122  1.219992e+02           tips
69   1.211936e+02       lochbuie
3    1.210675e+02        angelys
5    1.171669e+02        antares
92   9.629889e+01       pioppino
103  9.600250e+01      romanesco
34   8.390333e+01       cinnabar
102  6.898278e+01          rocha
87   6.661694e+01         pecans
65   6.006417e+01        kumquat
130  5.754917e+01             yu
98   4.800917e+01         purple
123  4.800167e+01           tree
31   4.111833e+01        chayote
39   3.847333e+01       cocktail
73   3.242194e+01          mache
76   3.034556e

In [78]:
q1hrs_cut = str(int(avg_memuse_byacct['used_hours'].quantile(.25)))
q2hrs_cut = str(int(avg_memuse_byacct['used_hours'].quantile(.50)))
q3hrs_cut = str(int(avg_memuse_byacct['used_hours'].quantile(.75)))
maxhrs = str(int(avg_memuse_byacct['used_hours'].max()))
print(q1hrs_cut + ' ' + q2hrs_cut + ' ' + q3hrs_cut + ' ' + maxhrs)
q1hrs_subplot_title = 'Low Time Usage - 0 hrs to ' + q1hrs_cut + ' hrs'
q2hrs_subplot_title = 'MedLo Time Usage - ' + q1hrs_cut + ' hrs to ' + q2hrs_cut + ' hrs'
q3hrs_subplot_title = 'MedHi Time Usage - ' + q2hrs_cut + ' hrs to ' + q3hrs_cut + ' hrs'
q4hrs_subplot_title = 'High Time Usage - ' + q3hrs_cut + ' hrs to ' + maxhrs + ' hrs'
q4hrs_subplot_title

207 2667 26197 3049115


'High Time Usage - 26197 hrs to 3049115 hrs'

In [81]:
mem_plot_avg = avg_memuse_byacct[['account','avg_usage','used_hours','mem_quartile', 'hours_quartile','total_time_ctgry']]#.set_index('account')
#mycolorscale = 'haline'

df=mem_plot_avg.sort_values(['avg_usage'], ascending=False)
fig = make_subplots(
                    rows=2, cols=2,
                    subplot_titles = (q1_subplot_title, q2_subplot_title, q3_subplot_title, q4_subplot_title),
                    vertical_spacing = 0.1,
                    horizontal_spacing = 0.2                    
                   )

fig.add_trace(go.Bar(y=df.loc[df['mem_quartile']=='Q1','account'], x=df.loc[df['mem_quartile']=='Q1','avg_usage'] 
                     ,orientation='h' 
                     ,hovertext=df['total_time_ctgry']
                     #,marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     , name='Low'), row=1, col=1)
fig.add_trace(go.Bar(y=df.loc[df['mem_quartile']=='Q2','account'], x=df.loc[df['mem_quartile']=='Q2','avg_usage']
                     ,orientation='h'
                     ,hovertext=df['total_time_ctgry']
                     #, marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     , name='MedLo'), row=1, col=2)
fig.add_trace(go.Bar(y=df.loc[df['mem_quartile']=='Q3','account'], x=df.loc[df['mem_quartile']=='Q3','avg_usage']
                     ,orientation='h'
                     ,hovertext=df['total_time_ctgry']
                     #, marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     , name='MedHi'), row=2, col=1)
fig.add_trace(go.Bar(y=df.loc[df['mem_quartile']=='Q4','account'], x=df.loc[df['mem_quartile']=='Q4','avg_usage']
                     ,orientation='h'
                     ,hovertext=df['total_time_ctgry']
                     #, marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     ,name='High'), row=2, col=2)

#fig.update_yaxes(type='category')
fig.update_xaxes(title_text = 'Mem % Usage')
fig.update_xaxes(range=[q1cut-.5, q2cut+2], row=1, col=2)
fig.update_xaxes(range=[q2cut-.5,q3cut+2], row=2, col=1)
fig.update_xaxes(range=[q3cut-.5,maxuse+2], row=2, col=2)


fig.update_layout(
                    title_text = 'ACCRE Accounts: Memory Usage as Percent of Memory Requested',             
                    showlegend=False,
                    height=1000#,
                    #coloraxis=dict(colorscale=mycolorscale, colorbar=dict(dtick=1, title='Total Time Quartile', titleside='right'))
                    )
#fig.show()
#marker=dict(color=df['hours_quartile'], colorscale=mycolorscale), 

In [82]:
mem_plot_avgT = avg_memuse_byacct[['account','avg_usage','used_hours','mem_quartile', 'hours_quartile']]#.set_index('account')
#mycolorscale = 'haline'

df=mem_plot_avg.sort_values(['avg_usage'], ascending=False)
figT = make_subplots(
                    rows=2, cols=2,
                    subplot_titles = (q1hrs_subplot_title, q2hrs_subplot_title, q3hrs_subplot_title, q4hrs_subplot_title),
                    vertical_spacing = 0.1,
                    horizontal_spacing = 0.2                    
                   )

figT.add_trace(go.Bar(y=df.loc[df['hours_quartile']=='Q1','account'], x=df.loc[df['hours_quartile']=='Q1','avg_usage'] 
                     ,orientation='h' 
                     #,marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     , name='Low'), row=1, col=1)
figT.add_trace(go.Bar(y=df.loc[df['hours_quartile']=='Q2','account'], x=df.loc[df['hours_quartile']=='Q2','avg_usage']
                     ,orientation='h'
                     #, marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     , name='MedLo'), row=1, col=2)
figT.add_trace(go.Bar(y=df.loc[df['hours_quartile']=='Q3','account'], x=df.loc[df['hours_quartile']=='Q3','avg_usage']
                     ,orientation='h'
                     #, marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     , name='MedHi'), row=2, col=1)
figT.add_trace(go.Bar(y=df.loc[df['hours_quartile']=='Q4','account'], x=df.loc[df['hours_quartile']=='Q4','avg_usage']
                     ,orientation='h'
                     #, marker=dict(color=df['hours_quartile'], coloraxis='coloraxis')
                     ,name='High'), row=2, col=2)

#figT.update_yaxes(type='category')
figT.update_xaxes(title_text = 'Mem % Usage')
#fig.update_xaxes(range=[q1cut-.5, q2cut+2], row=1, col=2)
#fig.update_xaxes(range=[q2cut-.5,q3cut+2], row=2, col=1)
#fig.update_xaxes(range=[q3cut-.5,max+2], row=2, col=2)

figT.update_layout(
                    title_text = 'ACCRE Accounts: Memory Usage as Percent of Memory Requested',             
                    showlegend=False,
                    height=1000#,
                    #coloraxis=dict(colorscale=mycolorscale, colorbar=dict(dtick=1, title='Total Time Quartile', titleside='right'))
                    )
#figT.show()

In [83]:
fig.show()

In [84]:
figT.show()

In [121]:
data_table1 = avg_memuse_byacct[avg_memuse_byacct['ttl_time_usg_ctgry'].isin(['1Low'])] \
    [['ttl_time_usg_ctgry', 'avg_mem_usg_ctgry', 'account', 'avg_usage', 'used_hours']] \
    .sort_values(['ttl_time_usg_ctgry', 'avg_usage'])
data_table1['used_hours'] = data_table1['used_hours'].astype(int)
data_table1.columns = ['Time Use Quartile', 'Memory Use % Quartile','Account', 'Avg Memory Use %', 'Total Hours Used' ]
#data_table1

In [122]:
data_table1b = avg_memuse_byacct[avg_memuse_byacct['ttl_time_usg_ctgry'].isin(['2MedLo'])] \
    [['ttl_time_usg_ctgry', 'avg_mem_usg_ctgry', 'account', 'avg_usage', 'used_hours']] \
    .sort_values(['ttl_time_usg_ctgry', 'avg_usage'])
data_table1b['used_hours'] = data_table1b['used_hours'].astype(int)
data_table1b.columns = ['Time Use Quartile', 'Memory Use % Quartile','Account', 'Avg Memory Use %', 'Total Hours Used' ]
#data_table1

In [123]:
data_table2 = avg_memuse_byacct[avg_memuse_byacct['ttl_time_usg_ctgry'].isin(['3MedHi'])] \
    [['ttl_time_usg_ctgry', 'avg_mem_usg_ctgry', 'account', 'avg_usage', 'used_hours']] \
    .sort_values(['ttl_time_usg_ctgry', 'avg_usage'])
data_table2['used_hours'] = data_table2['used_hours'].astype(int)
data_table2.columns = ['Time Use Quartile', 'Memory Use % Quartile','Account', 'Avg Memory Use %', 'Total Hours Used' ]
#data_table2

In [124]:
data_table2b = avg_memuse_byacct[avg_memuse_byacct['ttl_time_usg_ctgry'].isin(['4High'])] \
    [['ttl_time_usg_ctgry', 'avg_mem_usg_ctgry', 'account', 'avg_usage', 'used_hours']] \
    .sort_values(['ttl_time_usg_ctgry', 'avg_usage'])
data_table2b['used_hours'] = data_table2b['used_hours'].astype(int)
data_table2b.columns = ['Time Use Quartile', 'Memory Use % Quartile','Account', 'Avg Memory Use %', 'Total Hours Used' ]
#data_table2

In [132]:
figtbl1 = go.Figure(data = [go.Table(
    header=dict(values=list(data_table1.columns),
                               fill_color = 'paleturquoise',
                               align = 'left'),
    cells=dict(values=[data_table1['Time Use Quartile'], data_table1['Memory Use % Quartile'],
                       data_table1['Account'], data_table1['Avg Memory Use %'], data_table1['Total Hours Used']],
                                fill_color='lightgray',
                                 align='left')
)])

figtbl1.update_layout(title_text = 'Memory Usage as % of Memory Requested - Quartile 1 Total Hours')
figtbl1.show()

In [133]:
figtbl1b = go.Figure(data = [go.Table(
    header=dict(values=list(data_table1b.columns),
                               fill_color = 'paleturquoise',
                               align = 'left'),
    cells=dict(values=[data_table1b['Time Use Quartile'], data_table1b['Memory Use % Quartile'],
                       data_table1b['Account'], data_table1b['Avg Memory Use %'], data_table1b['Total Hours Used']],
                                fill_color='lightgray',
                                 align='left')
)])

figtbl1b.update_layout(title_text = 'Memory Usage as % of Memory Requested - Quartile 2 Total Hours')
figtbl1b.show()

In [134]:
figtbl2 = go.Figure(data = [go.Table(
    header=dict(values=list(data_table2.columns),
                               fill_color = 'paleturquoise',
                               align = 'left'),
    cells=dict(values=[data_table2['Time Use Quartile'], data_table2['Memory Use % Quartile'],
                       data_table2['Account'], data_table2['Avg Memory Use %'], data_table2['Total Hours Used']],
                                fill_color='lightgray',
                                 align='left')
)])

figtbl2.update_layout(title_text = 'Memory Usage as % of Memory Requested - Quartile 3 Total Hours')
figtbl2.show()

In [135]:
figtbl2b = go.Figure(data = [go.Table(
    header=dict(values=list(data_table2b.columns),
                               fill_color = 'paleturquoise',
                               align = 'left'),
    cells=dict(values=[data_table2b['Time Use Quartile'], data_table2b['Memory Use % Quartile'],
                       data_table2b['Account'], data_table2b['Avg Memory Use %'], data_table2b['Total Hours Used']],
                                fill_color='lightgray',
                                 align='left')
)])

figtbl2b.update_layout(title_text = 'Memory Usage as % of Memory Requested - Quartile 4 Total Hours')
figtbl2b.show()

In [136]:
fig.write_html('../notebooks/MemUse1.html', include_plotlyjs=True)
figT.write_html('../notebooks/MemUse2.html', include_plotlyjs=True)
figtbl1.write_html('../notebooks/MemUse3.html', include_plotlyjs=True)
figtbl1b.write_html('../notebooks/MemUse4.html', include_plotlyjs=True)
figtbl2.write_html('../notebooks/MemUse5.html', include_plotlyjs=True)
figtbl2.write_html('../notebooks/MemUse6.html', include_plotlyjs=True)

In [137]:
data_table1.to_csv('../Data/dt1.csv')
data_table1b.to_csv('../Data/dt1b.csv')
data_table2.to_csv('../Data/dt2.csv')
data_table2b.to_csv('../Data/dt2b.csv')

Data Transformations & Calculations
1.	Requested Memory Transformation
    a.	‘Mn’ entries = reqmem / (cpus / nodes)
    b.	All others = none
2.	Used Memory Transformation
    a.	All entries = usedmem / (cpus / nodes)
3.	Entries removed
    a.	State NOT ‘COMPLETED’
    b.	Partition NOT ‘production’
    c.	Exitcode NOT ‘0:0’
    d.	Usedmem = 0
    e.	Reqmem = 0
    f.	Cpus = 0
    g.	Cpus < nodes
4.	Memory Usage % Calculation = 
    Sum of requested memory for account / sum of used memory for account * 100
5.	Memory Usage % Quartiles
    min = 0.18%
    25 percentile = 8.02%
    50 percentile = 19.24%
    75 percentile = 34.34%
    max = 68.93%
6.	Time Usage in Hours Quartiles
    min = .033 hours
    25 percentile = 207.89 hours
    50 percentile = 2,667.72 hours
    75 percentile = 26,197.58 hours
    max = 3,049,115.97 hours



# Unused cells below

wavg_memuse_byacct = (memory_use.groupby('account')['mem_weight'].sum() / \
    memory_use.groupby('account')['used_total_seconds'].sum()).reset_index()
wavg_memuse_byacct.columns = ('account','wavg_usage')

wavg_memuse_byacct['quartile'] = pd.qcut(wavg_memuse_byacct['wavg_usage'], q = 4, labels=['1', '2','3','4'])

wavg_memuse_byacct.head()

wavg_memuse_byacct.sort_values('wavg_usage', ascending=False)

mem_plot_q1wavg = wavg_memuse_byacct[wavg_memuse_byacct['quartile'] == '1'][['account','wavg_usage']].set_index('account')
px_dfq1w=mem_plot_q1wavg.sort_values(['wavg_usage'])
mem_plot_q2wavg = wavg_memuse_byacct[wavg_memuse_byacct['quartile'] == '2'][['account','wavg_usage']].set_index('account')
px_dfq2w=mem_plot_q2wavg.sort_values(['wavg_usage'])
mem_plot_q3wavg = wavg_memuse_byacct[wavg_memuse_byacct['quartile'] == '3'][['account','wavg_usage']].set_index('account')
px_dfq3w=mem_plot_q3wavg.sort_values(['wavg_usage'])
mem_plot_q4wavg = wavg_memuse_byacct[wavg_memuse_byacct['quartile'] == '4'][['account','wavg_usage']].set_index('account')
px_dfq4w=mem_plot_q4wavg.sort_values(['wavg_usage'])


mem_plot_wtdavg = weightedavg_memuse_byacct[['account','weightedavg_usage']].set_index('account')
px_dfw=mem_plot_wtdavg.sort_values(['weightedavg_usage'])
px.bar(px_dfw, orientation='h', height = 2000 )

#slice out some short jobs
job_duration_short = memory_use[memory_use['used_hours'] <= 2]
job_duration_short.head().to_csv('../Data/short.csv')

#slice out some longer jobs
job_duration_long = memory_use[(memory_use['used_hours'] >4) & 
                               (memory_use['used_hours'] <8)
                              ]
job_duration_long.head().to_csv('../Data/longer.csv')

#slice out some really long jobs
job_duration_xlong = memory_use[(memory_use['used_hours'] > 72)
                              ]
job_duration_xlong.head().to_csv('../Data/xlong.csv')

#total hours used, for weighting
#ttl_used_hours = mem_use_by_acct['used_hours'].sum()

#add a column for group weighting by used_hours
#mem_use_by_acct['pct_ttl_used_hrs'] = mem_use_by_acct['used_hours'] / ttl_used_hours

#check that the total is 1
#mem_use_by_acct.pct_ttl_used_hrs.sum()

#add a column for weighted avg_usage
#mem_use_by_acct['weighted_avg_usage'] = mem_use_by_acct['avg_usage'] * mem_use_by_acct['pct_ttl_used_hrs']